demos/http/http_demo_s3_upload/http_demo_s3_upload.c

```diff
@@ -0,0 +1,624 @@
+                    serverHost,
+                    HTTPClient_strerror( httpStatus ) ) );
+    }
+
+    return httpStatus;
+}
+
+static HTTPStatus_t uploadS3ObjectFile( const TransportInterface_t * pTransportInterface,
+                                        const char * pPath )
+{
+    HTTPStatus_t httpStatus = HTTP_SUCCESS;
+
+    /* Initialize the request object. */
+    requestInfo.pHost = serverHost;
+    requestInfo.hostLen = serverHostLength;
+    requestInfo.method = HTTP_METHOD_PUT;
+    requestInfo.methodLen = HTTP_METHOD_PUT_LENGTH;
+    requestInfo.pPath = pPath;
+    requestInfo.pathLen = strlen( pPath );
+
+    /* Set "Connection" HTTP header to "keep-alive" so that multiple requests
+     * can be sent over the same established TCP connection. */
+    requestInfo.reqFlags = HTTP_REQUEST_KEEP_ALIVE_FLAG;
+
+    /* Set the buffer used for storing request headers. */
+    requestHeaders.pBuffer = userBuffer;
+    requestHeaders.bufferLen = USER_BUFFER_LENGTH;
+
+    /* Initialize the response object. The same buffer used for storing
+     * request headers is reused here. */
+    response.pBuffer = userBuffer;
+    response.bufferLen = USER_BUFFER_LENGTH;
+
+    if( httpStatus == HTTP_SUCCESS )
+    {
+        httpStatus = HTTPClient_InitializeRequestHeaders( &requestHeaders,
+                                                          &requestInfo );
+    }
+
+    if( httpStatus == HTTP_SUCCESS )
+    {
+        LogInfo( ( "Uploading file..." ) );
+        LogInfo( ( "Request Headers:\n%.*s",
+                   ( int32_t ) requestHeaders.headersLen,
+                   ( char * ) requestHeaders.pBuffer ) );
+        httpStatus = HTTPClient_Send( pTransportInterface,
+                                      &requestHeaders,
+                                      ( const uint8_t * ) DEMO_HTTP_UPLOAD_DATA,
+                                      DEMO_HTTP_UPLOAD_DATA_LENGTH,
+                                      &response,
+                                      0 );
+    }
+    else
+    {
+        LogError( ( "Failed to initialize HTTP request headers: Error=%s.",
+                    HTTPClient_strerror( httpStatus ) ) );
+    }
+
+    if( httpStatus == HTTP_SUCCESS )
+    {
+        LogInfo( ( "Received HTTP response from %s%s...",
+                   serverHost, pPath ) );
+        LogInfo( ( "Response Headers:\n%.*s",
+                   ( int32_t ) response.headersLen,
+                   response.pHeaders ) );
+        LogInfo( ( "Response Status:\n%u",
+                   response.statusCode ) );
+        LogInfo( ( "Response Body:\n%.*s\n",
+                   ( int32_t ) response.bodyLen,
+                   response.pBody ) );
+    }
+    else
+    {
+        LogError( ( "An error occurred in uploading the file."
+                    "Failed to send HTTP PUT request to %s%s: Error=%s.",
+                    serverHost, pPath, HTTPClient_strerror( httpStatus ) ) );
+    }
+
+    return httpStatus;
```

Comment:
Can we return success only for a 200 response in the response.statusCode?

Reply:
I am suggesting this because this demo is targeted to S3 and its status codes are well known to us. We don't want to continue the demo if S3 doesn't accept the object.